Distributed

MaxBenChrist opened the pull request:
So we are getting close to a distributed version of tsfresh.

I refactored the modules, wrote docstrings, added unit tests and added a chapter in the docs that explains how to use the distributor opbjects.

So, what is missing? I think we should still increase the test coverage, for example we can add some tests for the localdaskdistributor.

I am happy about feedback what still needs to be tested

MaxBenChrist commented:
There is still a small problem with the format of the features that are returned by the two Dask Distributors. I will upload a fix with some unit tests tomorrow

MaxBenChrist commented:
> There is still a small problem with the format of the features that are returned by the two Dask Distributors. I will upload a fix with some unit tests tomorrow

Done

nils-braun commented on tsfresh/utilities/string_manipulation.py:
```diff
@@ -1,6 +1,13 @@
 import ast
 import numpy as np
 from six import string_types
+import ipaddress
```
this is not needed anymore. Please remove :-)

MaxBenChrist replied:
Eagle eyes, Done!

nils-braun commented on tsfresh/feature_extraction/extraction.py:
```diff
@@ -208,28 +207,31 @@ def _do_extraction(df, column_id, column_value, column_kind,
-        distributor = distributor_class(n_workers=n_jobs,
-                                        disable_progressbar=disable_progressbar,
-                                        progressbar_title="Feature Extraction")
+    if not isinstance(distributor, DistributorBaseClass):
```
do we really want to be that strict? If a user implements everything that is needed, it is fine to have his own base class (he is of course responsible on his own).
Because of duck typing, it is barely needed to check for a type in python....

MaxBenChrist replied:
I want to catch stupid errors (e.g. user forgetting to instantiate the distributor object) with that strict class check.

I do not think that a user will be able to implement a custom distributor without inheriting from the base class. To write the distributor properly, you really need to understand how tsfresh uses it.

Also, there are not too many cluster frameworks, so also the number of possible distributors is limited. 

All in all, I think this motivates the strict test instead of duck typing.

nils-braun commented on tsfresh/feature_extraction/extraction.py:
```diff
@@ -208,28 +207,31 @@ def _do_extraction(df, column_id, column_value, column_kind,
+
         if n_jobs == 0:
-            distributor_class = MapDistributor
+            distributor = MapDistributor(disable_progressbar=disable_progressbar,
```
you could summarize the arguments into a dict and fill it in in both cases, to reduce code duplication.

MaxBenChrist replied:
I think it is quite easy to read in its current state...

nils-braun commented on tsfresh/feature_extraction/extraction.py:
```diff
@@ -111,10 +110,10 @@ def extract_features(timeseries_container, default_fc_parameters=None,
              distributor. See the utilities/distribution.py for more information. Leave to None, if you want
              TSFresh to choose the best distributor.
-    :type distributor_class: class
+    :type distributor: class
```
why two descriptions?

MaxBenChrist replied:
? not sure I understand your comment here

nils-braun commented:
Just for not loosing any information here: you have "just" added tests and documentation (thanks for that!) and not changed anything on the implementation itself, correct?

MaxBenChrist commented:
> Just for not loosing any information here: you have "just" added tests and documentation (thanks for that!) and not changed anything on the implementation itself, correct?

No, I changed a few things. For example the result list from the Dask Distributors needed to be flattened. Also I changed how the Distributor class is called inside `extract_features`

nils-braun commented:
Ah sorry, I meant in the distributed class itself (distributed.py) :-)
Ok, I have seen your changes in feature_extraction :+1:

MaxBenChrist commented:
> Ah sorry, I meant in the distributed class itself (distributed.py) :-)
> Ok, I have seen your changes in feature_extraction 👍

Some changes that come to my mind:

I moved the implementation of one of the methods from base class to the child classes.
Also, the constructor of every child class has different arguments now. 
I made the calculate chunksize method non private
